lib/cache/memory-cache-store.js

```diff
@@ -0,0 +1,402 @@
+   */
+  createWriteStream (req, opts) {
+    if (typeof req !== 'object') {
+      throw new TypeError(`expected req to be object, got ${typeof req}`)
+    }
+    if (typeof opts !== 'object') {
+      throw new TypeError(`expected value to be object, got ${typeof opts}`)
+    }
+
+    if (this.isFull) {
+      return undefined
+    }
+
+    const values = this.#getValuesForRequest(req, true)
+
+    let value = this.#findValue(req, values)
+    if (!value) {
+      // The value doesn't already exist, meaning we haven't cached this
+      //  response before. Let's assign it a value and insert it into our data
+      //  property.
+
+      if (this.isFull) {
+        // Or not, we don't have space to add another response
+        return undefined
+      }
+
+      this.#entryCount++
+
+      value = {
+        readers: 0,
+        readLock: false,
+        writeLock: false,
+        opts,
+        body: []
+      }
+
+      // We want to sort our responses in decending order by their deleteAt
+      //  timestamps so that deleting expired responses is faster
+      if (
+        values.length === 0 ||
+        opts.deleteAt < values[values.length - 1].deleteAt
+      ) {
+        // Our value is either the only response for this path or our deleteAt
+        //  time is sooner than all the other responses
+        values.push(value)
+      } else if (opts.deleteAt >= values[0].deleteAt) {
+        // Our deleteAt is later than everyone elses
+        values.unshift(value)
+      } else {
+        // We're neither in the front or the end, let's just binary search to
+        //  find our stop we need to be in
+        let startIndex = 0
+        let endIndex = values.length
+        while (true) {
+          if (startIndex === endIndex) {
+            values.splice(startIndex, 0, value)
+            break
+          }
+
+          const middleIndex = Math.floor((startIndex + endIndex) / 2)
+          const middleValue = values[middleIndex]
+          if (opts.deleteAt === middleIndex) {
+            values.splice(middleIndex, 0, value)
+            break
+          } else if (opts.deleteAt > middleValue.opts.deleteAt) {
+            endIndex = middleIndex
+            continue
+          } else {
+            startIndex = middleIndex
+            continue
+          }
+        }
+      }
+    } else {
+      // Check if there's already another request writing to the value or
+      //  a request reading from it
+      if (value.writeLock || value.readLock) {
+        return undefined
+      }
+
+      // Empty it so we can overwrite it
+      value.body = []
+    }
+
+    const writable = new MemoryStoreWritableStream(
+      value,
+      this.#maxEntrySize
+    )
+
+    // Remove the value if there was some error
+    writable.on('error', () => {
+      values.filter(current => value !== current)
+    })
+
+    writable.on('bodyOversized', () => {
+      values.filter(current => value !== current)
+    })
```

Comment:
This is super weird.... What's going on?

Reply:
There was an error or the body turned out to be oversized. If we keep storing it, it would be in an incomplete state so any future requests that use that value would also be incomplete, so we need to delete it